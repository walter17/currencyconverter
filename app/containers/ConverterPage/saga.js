/**
 * Redux Saga
 * API Request handler for Currency Exchange Rates
 * -----------------------------------------------
 * Author: Walter Ramos
 * Last Modified: 02/28/2021
 */
import { call, put, takeLatest, select } from 'redux-saga/effects';
import {
  LOAD_RATELIST,
  CURRENCY_LIST_URI,
  CONVERT,
  CONVERT_URI,
} from 'containers/App/constants';
import {
  exchangeListSuccess,
  exchangeListError,
  convertSuccess,
  convertFailed,
} from 'containers/App/actions';

import {
  makeSelectConvertValue,
  makeSelectLeftCurrency,
  makeSelectRightCurrency,
} from 'containers/App/selectors';
import request from 'utils/request';

// Retrieve currency exchange rates
export function* getCurrencyList() {
  // Construct API headers
  const requestURL = `${CURRENCY_LIST_URI}?access_key=252f4905adac6f1633d32f17b8b6cba0&base=USD`;
  try {
    // Call our request helper (see 'utils/request')
    const results = yield call(request, requestURL);
    if (results.success) {
      yield put(exchangeListSuccess(results.rates));
    } else {
      yield put(exchangeListError(results));
    }
  } catch (err) {
    yield put(exchangeListError(err));
  }
}

// Process currency conversion
export function* convertCurrency() {
  // Retrieve
  const lCurrency = yield select(makeSelectLeftCurrency('convert'));
  const rCurrency = yield select(makeSelectRightCurrency('convert'));
  const convertVal = yield select(makeSelectConvertValue('convert'));

  // Construct API headers
  const requestURL = `${CONVERT_URI}?access_key=252f4905adac6f1633d32f17b8b6cba0&from=${lCurrency}&to=${rCurrency}&amount=${convertVal}`;
  try {
    // Call our request helper (see 'utils/request')
    const results = yield call(request, requestURL);
    if (results.success) {
      yield put(convertSuccess(results));
    } else {
      yield put(convertFailed(results));
    }
  } catch (err) {
    yield put(convertFailed(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* currencyListData() {
  // Watches for LOAD_RATELIST, CONVERT actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(LOAD_RATELIST, getCurrencyList);
  yield takeLatest(CONVERT, convertCurrency);
}
