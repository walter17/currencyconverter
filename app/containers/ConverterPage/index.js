/* eslint-disable react/prop-types */
/* eslint-disable react/no-unused-prop-types */
/**
 * Converter Page File
 */
import React, { useEffect, memo } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';

import { useInjectReducer } from 'utils/injectReducer';
import { useInjectSaga } from 'utils/injectSaga';
import {
  makeSelectCurrencyList,
  makeSelectLoading,
  makeSelectError,
  makeSelectConvertError,
  makeSelectConvertLoading,
  makeSelectConvertData,
} from 'containers/App/selectors';
import { loadExchangeList, sendToConvert } from 'containers/App/actions';
import { getCountryByCode } from 'containers/App/commons';
import {
  Box,
  Container,
  Grid,
  InputAdornment,
  LinearProgress,
  OutlinedInput,
  Paper,
  TextField,
  Typography,
} from '@material-ui/core';
import { Autocomplete } from '@material-ui/lab';
import reducer from 'containers/App/reducer';
import saga from './saga';

// Common Dependencies

const key = 'exchangerates';

export function ConverterPage({
  loading,
  error,
  list,
  loadList,
  startCurrencyConvert,
  convertedResult,
  convertError,
}) {
  useInjectReducer({ key, reducer });
  useInjectSaga({ key, saga });

  const [rates, setRates] = React.useState([]);
  const [loadingRates, setLoading] = React.useState(false);
  const [leftCurrency, setLeftCurrency] = React.useState(false);
  const [rightCurrency, setRightCurrency] = React.useState(false);
  const [leftFocus, setLeftFocus] = React.useState(false);
  const [rightFocus, setRightFocus] = React.useState(false);
  const [lockedPointers, setLocked] = React.useState(true);
  const [leftConvertedValue, setLeftConvertValue] = React.useState(0);
  const [rightConvertedValue, setRightConvertValue] = React.useState(0);

  useEffect(() => {
    if (list.length === 0 && !loadingRates) {
      setLoading(true);
      // Load list if empty
      loadList();
      return;
    }
    setRates(Object.keys(list).map(keyItem => [keyItem, list[keyItem]]));
  }, [list]);

  useEffect(() => {
    if (leftConvertedValue >= 1 && leftFocus && !lockedPointers) {
      setRightFocus(false);
      // Perform conversion calls
      startCurrencyConvert(leftCurrency, rightCurrency, leftConvertedValue);
    }
  }, [leftConvertedValue]);

  useEffect(() => {
    if (rightConvertedValue >= 1 && rightFocus && !lockedPointers) {
      setLeftFocus(false);
      // Perform conversion calls
      startCurrencyConvert(rightCurrency, leftCurrency, rightConvertedValue);
    }
  }, [rightConvertedValue]);

  useEffect(() => {
    if (!convertedResult.success) {
      setLocked(false); // UnLock next calls
      return; // Return | Break if failed
    }
    setLocked(true); // Lock next calls
    if (leftFocus) {
      setRightConvertValue(convertedResult.result);
    } else if (rightFocus) {
      setLeftConvertValue(convertedResult.result);
    } else {
      setRightConvertValue(0);
      setLeftConvertValue(0);
    }
  }, [convertedResult]);

  return (
    <Container maxWidth="xl">
      <Typography variant="h4">Currency converter</Typography>
      <Typography variant="body1">
        Please enter the amount to convert on any field
      </Typography>
      {loading && <LinearProgress />}

      {error ||
        (convertError && (
          <Typography variant="body2" color="secondary">
            Something went wrong. Please try again!
          </Typography>
        ))}
      <Grid
        container
        direction="row"
        justify="center"
        alignItems="center"
        spacing={2}
      >
        <Grid item xs={6}>
          <Paper
            variant="outlined"
            square
            elevation={4}
            style={{ padding: 40 }}
          >
            <ConverterForm
              rates={rates}
              setCurrencyValue={val => {
                setLeftConvertValue(val);
                setLeftFocus(true);
                setLocked(false); // UnLock next calls
              }}
              currValue={leftConvertedValue}
              setCurrencyType={type => {
                setLeftCurrency(type);
              }}
            />
          </Paper>
          <Typography variant="body1" style={{ margin: '20px 40px' }}>
            1 USD = 0.83
          </Typography>
        </Grid>
        <Grid item xs={6}>
          <Paper
            variant="outlined"
            square
            elevation={4}
            style={{ padding: 40 }}
          >
            <ConverterForm
              rates={rates}
              setCurrencyValue={val => {
                setRightConvertValue(val);
                setRightFocus(true);
                setLocked(false); // UnLock next calls
              }}
              currValue={rightConvertedValue}
              setCurrencyType={type => {
                setRightCurrency(type);
              }}
            />
          </Paper>
          <Typography variant="body1" style={{ margin: '20px 40px' }}>
            1 USD = 0.83
          </Typography>
        </Grid>
      </Grid>
    </Container>
  );
}

function ConverterForm({
  rates,
  setCurrencyValue,
  currValue,
  setCurrencyType,
}) {
  const handleValueChange = event => {
    setCurrencyValue(event.target.value);
  };

  const handleCurrencyChange = (event, value) => {
    if (event) {
      setCurrencyType(value[0]);
      return;
    }
    setCurrencyType(false);
  };

  return (
    <Box>
      <Typography variant="body2">Currency</Typography>
      <Autocomplete
        id="combo-box-demo"
        options={rates}
        getOptionLabel={option =>
          `${
            getCountryByCode(option[0])[0]
              ? getCountryByCode(option[0])[0].name
              : '-'
          } (${option[0]})`
        }
        style={{ margin: '20px auto' }}
        renderInput={params => <TextField {...params} variant="outlined" />}
        onChange={handleCurrencyChange}
      />

      <Typography variant="body2">Enter amount</Typography>
      <OutlinedInput
        id="outlined-basic"
        style={{ margin: '20px auto', width: '100%' }}
        endAdornment={<InputAdornment position="start">$</InputAdornment>}
        value={currValue}
        onChange={e => handleValueChange(e)}
      />
    </Box>
  );
}

ConverterPage.propTypes = {
  setCurrencyType: PropTypes.func,
  loading: PropTypes.bool,
  error: PropTypes.any,
  list: PropTypes.any,
  loadList: PropTypes.func,
  startCurrencyConvert: PropTypes.func,
  convertedResult: PropTypes.any,
  convertLoading: PropTypes.bool,
  convertError: PropTypes.any,
};

ConverterForm.propTypes = {
  rates: PropTypes.any,
  setCurrencyValue: PropTypes.func,
  currValue: PropTypes.any,
};

const mapStateToProps = createStructuredSelector({
  // currency rates list
  list: makeSelectCurrencyList(),
  loading: makeSelectLoading(),
  error: makeSelectError(),

  // converted currency props
  convertedResult: makeSelectConvertData('convert'),
  convertLoading: makeSelectConvertLoading('convert'),
  convertError: makeSelectConvertError('convert'),
});

export function mapDispatchToProps(dispatch) {
  return {
    loadList: () => {
      dispatch(loadExchangeList());
    },
    startCurrencyConvert: (leftCurrency, rightCurrency, valueToConvert) => {
      dispatch(sendToConvert(leftCurrency, rightCurrency, valueToConvert));
    },
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(ConverterPage);
