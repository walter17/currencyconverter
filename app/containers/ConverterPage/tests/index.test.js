import React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import { ConnectedRouter } from 'connected-react-router/immutable';
import { createMemoryHistory } from 'history';
import configureStore from '../../../configureStore';
import ConverterPage from '../index';

describe('<ConverterPage />', () => {
  const history = createMemoryHistory();
  const store = configureStore({}, history);

  it('should render a Converter Page Component', () => {
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <ConverterPage
              setCurrencyType={() => {}}
              loading={false}
              error={false}
              list={[]}
              loadList={() => {}}
              startCurrencyConvert={() => {}}
              convertedResult={[]}
              convertLoading={false}
              convertError={false}
            />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
