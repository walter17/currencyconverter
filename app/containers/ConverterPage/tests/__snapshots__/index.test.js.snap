// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConverterPage /> should render a Converter Page Component 1`] = `
<div
  class="MuiBox-root MuiBox-root-1"
>
  <h4
    class="MuiTypography-root MuiTypography-h4"
  >
    Currency converter
  </h4>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    Please enter the amount to convert on any field
  </p>
  <p
    class="MuiTypography-root MuiTypography-body2 MuiTypography-colorSecondary"
  >
    Something went wrong. Please try again!
  </p>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined"
        style="padding: 40px;"
      >
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <p
            class="MuiTypography-root MuiTypography-body2"
          >
            Currency
          </p>
          <div
            aria-expanded="false"
            class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
            role="combobox"
            style="margin: 20px auto;"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
              >
                <input
                  aria-autocomplete="list"
                  aria-invalid="false"
                  autocapitalize="none"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                  id="mui-66481"
                  spellcheck="false"
                  type="text"
                  value=""
                />
                <div
                  class="MuiAutocomplete-endAdornment"
                >
                  <button
                    aria-label="Clear"
                    class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                    tabindex="-1"
                    title="Clear"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-label="Open"
                    class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                    tabindex="-1"
                    title="Open"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                  style="padding-left: 8px;"
                >
                  <legend
                    class="PrivateNotchedOutline-legend-4"
                    style="width: 0.01px;"
                  >
                    <span>
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2"
          >
            Enter amount
          </p>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
            style="margin: 20px auto; width: 100%;"
          >
            <input
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              type="text"
              value="0"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
              >
                $
              </p>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-4"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1"
        style="margin: 20px 40px;"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined"
        style="padding: 40px;"
      >
        <div
          class="MuiBox-root MuiBox-root-7"
        >
          <p
            class="MuiTypography-root MuiTypography-body2"
          >
            Currency
          </p>
          <div
            aria-expanded="false"
            class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
            role="combobox"
            style="margin: 20px auto;"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
              >
                <input
                  aria-autocomplete="list"
                  aria-invalid="false"
                  autocapitalize="none"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                  id="mui-80433"
                  spellcheck="false"
                  type="text"
                  value=""
                />
                <div
                  class="MuiAutocomplete-endAdornment"
                >
                  <button
                    aria-label="Clear"
                    class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                    tabindex="-1"
                    title="Clear"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-label="Open"
                    class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                    tabindex="-1"
                    title="Open"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                  style="padding-left: 8px;"
                >
                  <legend
                    class="PrivateNotchedOutline-legend-4"
                    style="width: 0.01px;"
                  >
                    <span>
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2"
          >
            Enter amount
          </p>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
            style="margin: 20px auto; width: 100%;"
          >
            <input
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              type="text"
              value="0"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
              >
                $
              </p>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-4"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1"
        style="margin: 20px 40px;"
      />
    </div>
  </div>
</div>
`;
